File: backend/app/domain/billing.py
"""Billing DTOs (cycle 4)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

PlanTier = Literal["starter", "growth", "team", "enterprise"]
BillingStatus = Literal["trialing", "active", "past_due", "canceled", "incomplete"]


class CheckoutRequest(BaseModel):
    """POST /api/billing/checkout payload."""

    model_config = ConfigDict(extra="forbid")

    plan: PlanTier = Field(description="Plan to upgrade to (e.g. 'growth').")


class CheckoutResponse(BaseModel):
    """Returned by POST /api/billing/checkout and /portal."""

    model_config = ConfigDict(extra="ignore")

    url: str
    session_id: str


class BillingStatusOut(BaseModel):
    """Returned by GET /api/billing/status."""

    model_config = ConfigDict(extra="ignore")

    plan: PlanTier
    status: BillingStatus
    seats_used: int
    seats_limit: int
    properties_used: int
    properties_limit: int
    ai_listings_used_month: int
    ai_listings_limit_month: int
    current_period_end: datetime | None = None
    cancel_at_period_end: bool = False
    trial_ends_at: datetime | None = None
    is_paid: bool = False
    is_within_plan_limits: bool = True


class WebhookEventIn(BaseModel):
    """Raw payload of POST /api/billing/webhook (real webhook)."""

    model_config = ConfigDict(extra="ignore")

    id: str
    type: str
    data: dict[str, Any]


class WebhookAck(BaseModel):
    """Returned by POST /api/billing/webhook."""

    status: str
    received: int
    processed: int

File: web/lib/billing.ts
/**
 * Typed billing API client (cycle 4).
 */

import { apiGet, apiPost, apiPostNoBody, apiDelete } from "./api";

export type PlanTier = "starter" | "growth" | "team";
export type BillingStatus =
  | "trialing"
  | "active"
  | "past_due"
  | "canceled"
  | "incomplete";

export interface BillingStatusOut {
  plan: PlanTier;
  status: BillingStatus;
  seats_used: number;
  seats_limit: number;
  properties_used: number;
  properties_limit: number;
  ai_listings_used_month: number;
  ai_listings_limit_month: number;
  current_period_end?: string | null;
  cancel_at_period_end: boolean;
  trial_ends_at?: string | null;
  is_paid: boolean;
  is_within_plan_limits: boolean;
}

export interface CheckoutResponse {
  url: string;
  session_id: string;
}

export const PLAN_PRICING: Record<
  Exclude<PlanTier, "starter">,
  { name: string; price: number; seats: number; properties: number; ai: number }
> = {
  growth: { name: "Growth", price: 29, seats: 3, properties: 25, ai: 200 },
  team: { name: "Team", price: 99, seats: 10, properties: 100, ai: 1000 },
};

export const billingApi = {
  status: () => apiGet<BillingStatusOut>("/api/billing/status"),
  checkout: (plan: PlanTier) =>
    apiPost<CheckoutResponse>("/api/billing/checkout", { plan }),
  portal: () => apiPost<CheckoutResponse>("/api/billing/portal", {}),
  cancel: (subscriptionId: string) =>
    apiDelete<void>(`/api/billing/subscriptions/${subscriptionId}`),
  keepAlive: () => apiPostNoBody<void>("/api/billing/heartbeat"),
};
